In [55]:
# Importing necessary packages
import numpy as np 
import polars as pl 
import pandas as pd

In [56]:
# Reading in the data
df = pl.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc.csv")

In [57]:
# Taking a high level look at the data using head
df.head()

shape: (5, 144)
┌─────────────┬─────────────┬────────────┬────────────┬───┬───────────┬────────────┬───────┬───────┐
│ R_fighter   ┆ B_fighter   ┆ Referee    ┆ date       ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age ┆ R_age │
│ ---         ┆ ---         ┆ ---        ┆ ---        ┆   ┆ ms        ┆ bs         ┆ ---   ┆ ---   │
│ str         ┆ str         ┆ str        ┆ str        ┆   ┆ ---       ┆ ---        ┆ f64   ┆ f64   │
│             ┆             ┆            ┆            ┆   ┆ f64       ┆ f64        ┆       ┆       │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪═══════════╪════════════╪═══════╪═══════╡
│ Adrian      ┆ Gustavo     ┆ Chris      ┆ 2021-03-20 ┆ … ┆ 177.8     ┆ 135.0      ┆ 31.0  ┆ 27.0  │
│ Yanez       ┆ Lopez       ┆ Tognoni    ┆            ┆   ┆           ┆            ┆       ┆       │
│ Trevin      ┆ Roman       ┆ Herb Dean  ┆ 2021-03-20 ┆ … ┆ 187.96    ┆ 185.0      ┆ 32.0  ┆ 28.0  │
│ Giles       ┆ Dolidze     ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Tai Tuivasa ┆ Harry       ┆ Herb Dean  ┆ 2021-03-20 ┆ … ┆ 190.5     ┆ 264.0      ┆ 32.0  ┆ 28.0  │
│             ┆ Hunsucker   ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Cheyanne    ┆ Montserrat  ┆ Mark Smith ┆ 2021-03-20 ┆ … ┆ 160.02    ┆ 115.0      ┆ 28.0  ┆ 25.0  │
│ Buys        ┆ Conejo      ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Marion      ┆ Macy        ┆ Mark Smith ┆ 2021-03-20 ┆ … ┆ 172.72    ┆ 135.0      ┆ 29.0  ┆ 43.0  │
│ Reneau      ┆ Chiasson    ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
└─────────────┴─────────────┴────────────┴────────────┴───┴───────────┴────────────┴───────┴───────┘

We looked at the first 5 rows(or 5 latest fights) of each of the 144 columns here.

In [58]:
# Taking a high level look at the data using tail
df.tail()

shape: (5, 144)
┌─────────────┬─────────────┬────────────┬────────────┬───┬───────────┬────────────┬───────┬───────┐
│ R_fighter   ┆ B_fighter   ┆ Referee    ┆ date       ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age ┆ R_age │
│ ---         ┆ ---         ┆ ---        ┆ ---        ┆   ┆ ms        ┆ bs         ┆ ---   ┆ ---   │
│ str         ┆ str         ┆ str        ┆ str        ┆   ┆ ---       ┆ ---        ┆ f64   ┆ f64   │
│             ┆             ┆            ┆            ┆   ┆ f64       ┆ f64        ┆       ┆       │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪═══════════╪════════════╪═══════╪═══════╡
│ Remco       ┆ Alberta     ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 260.0      ┆ null  ┆ null  │
│ Pardoel     ┆ Cerra Leon  ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Orlando     ┆ Robert      ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 170.0      ┆ null  ┆ null  │
│ Wiet        ┆ Lucarelli   ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Johnny      ┆ David       ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 210.0      ┆ null  ┆ null  │
│ Rhodes      ┆ Levicki     ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Patrick     ┆ Ray Wizard  ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 225.0      ┆ null  ┆ 30.0  │
│ Smith       ┆             ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Scott       ┆ Sean        ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 210.0      ┆ 18.0  ┆ null  │
│ Morris      ┆ Daugherty   ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
└─────────────┴─────────────┴────────────┴────────────┴───┴───────────┴────────────┴───────┴───────┘

We looked at the last 5 rows(or first 5 fights) of each of the 144 columns here.

In [59]:
# Looking at and understanding the names/meanings of the columns
df.columns

['R_fighter',
 'B_fighter',
 'Referee',
 'date',
 'location',
 'Winner',
 'title_bout',
 'weight_class',
 'B_avg_KD',
 'B_avg_opp_KD',
 'B_avg_SIG_STR_pct',
 'B_avg_opp_SIG_STR_pct',
 'B_avg_TD_pct',
 'B_avg_opp_TD_pct',
 'B_avg_SUB_ATT',
 'B_avg_opp_SUB_ATT',
 'B_avg_REV',
 'B_avg_opp_REV',
 'B_avg_SIG_STR_att',
 'B_avg_SIG_STR_landed',
 'B_avg_opp_SIG_STR_att',
 'B_avg_opp_SIG_STR_landed',
 'B_avg_TOTAL_STR_att',
 'B_avg_TOTAL_STR_landed',
 'B_avg_opp_TOTAL_STR_att',
 'B_avg_opp_TOTAL_STR_landed',
 'B_avg_TD_att',
 'B_avg_TD_landed',
 'B_avg_opp_TD_att',
 'B_avg_opp_TD_landed',
 'B_avg_HEAD_att',
 'B_avg_HEAD_landed',
 'B_avg_opp_HEAD_att',
 'B_avg_opp_HEAD_landed',
 'B_avg_BODY_att',
 'B_avg_BODY_landed',
 'B_avg_opp_BODY_att',
 'B_avg_opp_BODY_landed',
 'B_avg_LEG_att',
 'B_avg_LEG_landed',
 'B_avg_opp_LEG_att',
 'B_avg_opp_LEG_landed',
 'B_avg_DISTANCE_att',
 'B_avg_DISTANCE_landed',
 'B_avg_opp_DISTANCE_att',
 'B_avg_opp_DISTANCE_landed',
 'B_avg_CLINCH_att',
 'B_avg_CLINCH_lande

The column names are shortened but record 
- Fighter Names(Blue and Red Corner)
- Generic information: Date, location, referee name, whether the fight was for a title and the weight class
- Blue Fighter average fighting career statistics for/against: knockdowns, significant strikes percentage/attempted/landed, takedown percentage/attempted/landed, submissions, reversals, total strikes attempted/landed, head/body/leg strikes attempted/landed, attacks landed/attempted at clinch/distance/ground ranges, control time for/against
- Blue fighter generic career statistics: Total time/rounds fought, title bouts, win/lose streak, longest winning streak, total wins/losses/draws, wins by majority/split/unanimous decision, wins by ko/tko, wins by submission, wins by doctor stoppage
- Blue fighter physiological statistics: Stance, Height(cms), Weight, Reach
- Red Fighter average fighting career statistics for/against: knockdowns, significant strikes percentage/attempted/landed, takedown percentage/attempted/landed, submissions, reversals, total strikes attempted/landed, head/body/leg strikes attempted/landed, attacks landed/attempted at clinch/distance/ground ranges, control time for/against
- Red fighter generic career statistics: Total time/rounds fought, title bouts, win/lose streak, longest winning streak, total wins/losses/draws, wins by majority/split/unanimous decision, wins by ko/tko, wins by submission, wins by doctor stoppage
- Red fighter physiological statistics: Stance, Height(cms), Weight, Reach
Since this is a fairly new sport, some non-obvious but important terms need to be clarified for ease of understanding to non-fight fans, here are the definitions:-
> Knockout (KO): When an athlete is knocked unconscious due to strikes or other impact. A knockout is not a technical knockout, but rather one that leaves the fighter physically unable to continue.
> Technical Knockout (TKO): Abbreviation for technical knockout, when an athlete is unable to continue, usually due to injury.
> Submission: When an athlete taps out or verbally concedes the match due to pain, to avoid injury, to avoid being choked or due to a desire to end the match. A referee has the power to force a fighter to submit at his discretion due to risk of severe injury/loss of consciousness/death.
> Knockdown (KD): When a fighter touches the floor of the ring with any part of the body other than the feet following a hit, but is able to rise back up and continue fighting
> Reversal: When an athlete moves from an inferior position into a superior position.
> Unanimous Decision: All judges scored the fight in favor of one fighter.
> Majority Decision: 2 judges scored the fight in favor of the winner while the third scored a draw.
> Split Decision: 2 judges scored one corner as the winner while the third voted for the opposing corner.
> Majority decision draw: Two judges voted for a draw while one judge voted for a winner.
> Split decision draw: The three judges scored the contest as blue corner win, red corner win and draw.

In [67]:
#Counting the number of times a value appears in each column
for x in df:
    y = x.value_counts().sort(by = 'counts', descending = True)
    print(y)

shape: (1_514, 2)
┌──────────────────────────┬────────┐
│ R_fighter                ┆ counts │
│ ---                      ┆ ---    │
│ str                      ┆ u32    │
╞══════════════════════════╪════════╡
│ Jim Miller               ┆ 27     │
│ Demian Maia              ┆ 24     │
│ Diego Sanchez            ┆ 24     │
│ Donald Cerrone           ┆ 24     │
│ …                        ┆ …      │
│ DeAnna Bennett           ┆ 1      │
│ Kyle Daukaus             ┆ 1      │
│ Rameau Thierry Sokoudjou ┆ 1      │
│ Claude Patrick           ┆ 1      │
└──────────────────────────┴────────┘
shape: (1_987, 2)
┌──────────────────┬────────┐
│ B_fighter        ┆ counts │
│ ---              ┆ ---    │
│ str              ┆ u32    │
╞══════════════════╪════════╡
│ Jeremy Stephens  ┆ 21     │
│ Charles Oliveira ┆ 19     │
│ Nik Lentz        ┆ 15     │
│ Rafael Dos Anjos ┆ 14     │
│ …                ┆ …      │
│ John Dowdy       ┆ 1      │
│ Igor Araujo      ┆ 1      │
│ Max Rohskopf     ┆ 1      │
│ Jo

In [66]:
#Taking a look at the shape of the data
df.shape

(6012, 144)

This reveals there have been 6012 fights and 144 variables have been tracked in the columns. The column names and what they mean have been described above.

In [73]:
# Reading in the data using pandas
dfpandas = pd.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc.csv")

In [97]:
# Looking at the data structure
dfpandas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6012 entries, 0 to 6011
Columns: 144 entries, R_fighter to R_age
dtypes: bool(1), float64(106), int64(28), object(9)
memory usage: 6.6+ MB


In [109]:
# Printing the names and types of the data
for x in dfpandas:
    print("Name(Type) : " + x + "(" + str(type(x)) + ")")

Name(Type) : R_fighter(<class 'str'>)
Name(Type) : B_fighter(<class 'str'>)
Name(Type) : Referee(<class 'str'>)
Name(Type) : date(<class 'str'>)
Name(Type) : location(<class 'str'>)
Name(Type) : Winner(<class 'str'>)
Name(Type) : title_bout(<class 'str'>)
Name(Type) : weight_class(<class 'str'>)
Name(Type) : B_avg_KD(<class 'str'>)
Name(Type) : B_avg_opp_KD(<class 'str'>)
Name(Type) : B_avg_SIG_STR_pct(<class 'str'>)
Name(Type) : B_avg_opp_SIG_STR_pct(<class 'str'>)
Name(Type) : B_avg_TD_pct(<class 'str'>)
Name(Type) : B_avg_opp_TD_pct(<class 'str'>)
Name(Type) : B_avg_SUB_ATT(<class 'str'>)
Name(Type) : B_avg_opp_SUB_ATT(<class 'str'>)
Name(Type) : B_avg_REV(<class 'str'>)
Name(Type) : B_avg_opp_REV(<class 'str'>)
Name(Type) : B_avg_SIG_STR_att(<class 'str'>)
Name(Type) : B_avg_SIG_STR_landed(<class 'str'>)
Name(Type) : B_avg_opp_SIG_STR_att(<class 'str'>)
Name(Type) : B_avg_opp_SIG_STR_landed(<class 'str'>)
Name(Type) : B_avg_TOTAL_STR_att(<class 'str'>)
Name(Type) : B_avg_TOTAL_STR_

In [111]:
# Looking at summary statistics
df.describe()

shape: (9, 145)
┌───────────┬─────────┬───────────┬───────────┬───┬───────────┬────────────┬───────────┬───────────┐
│ describe  ┆ R_fight ┆ B_fighter ┆ Referee   ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age     ┆ R_age     │
│ ---       ┆ er      ┆ ---       ┆ ---       ┆   ┆ ms        ┆ bs         ┆ ---       ┆ ---       │
│ str       ┆ ---     ┆ str       ┆ str       ┆   ┆ ---       ┆ ---        ┆ f64       ┆ f64       │
│           ┆ str     ┆           ┆           ┆   ┆ f64       ┆ f64        ┆           ┆           │
╞═══════════╪═════════╪═══════════╪═══════════╪═══╪═══════════╪════════════╪═══════════╪═══════════╡
│ count     ┆ 6012    ┆ 6012      ┆ 6012      ┆ … ┆ 6012.0    ┆ 6012.0     ┆ 6012.0    ┆ 6012.0    │
│ null_coun ┆ 0       ┆ 0         ┆ 32        ┆ … ┆ 406.0     ┆ 2.0        ┆ 172.0     ┆ 63.0      │
│ t         ┆         ┆           ┆           ┆   ┆           ┆            ┆           ┆           │
│ mean      ┆ null    ┆ null      ┆ null      ┆ … ┆ 183.56234 ┆ 170.686356 ┆ 29.293151 ┆ 29.598252 │
│           ┆         ┆           ┆           ┆   ┆ 7         ┆            ┆           ┆           │
│ std       ┆ null    ┆ null      ┆ null      ┆ … ┆ 10.585313 ┆ 35.724626  ┆ 4.063297  ┆ 4.167755  │
│ min       ┆ Aalon   ┆ Aalon     ┆ Adam      ┆ … ┆ 152.4     ┆ 115.0      ┆ 18.0      ┆ 19.0      │
│           ┆ Cruz    ┆ Cruz      ┆ Cheadle   ┆   ┆           ┆            ┆           ┆           │
│ max       ┆ Zubaira ┆ Zubaira   ┆ Yves      ┆ … ┆ 213.36    ┆ 345.0      ┆ 51.0      ┆ 47.0      │
│           ┆ Tukhugo ┆ Tukhugov  ┆ Lavigne   ┆   ┆           ┆            ┆           ┆           │
│           ┆ v       ┆           ┆           ┆   ┆           ┆            ┆           ┆           │
│ median    ┆ null    ┆ null      ┆ null      ┆ … ┆ 182.88    ┆ 170.0      ┆ 29.0      ┆ 29.0      │
│ 25%       ┆ null    ┆ null      ┆ null      ┆ … ┆ 177.8     ┆ 145.0      ┆ 26.0      ┆ 27.0      │
│ 75%       ┆ null    ┆ null      ┆ null      ┆ … ┆ 190.5     ┆ 185.0      ┆ 32.0      ┆ 32.0      │
└───────────┴─────────┴───────────┴───────────┴───┴───────────┴────────────┴───────────┴───────────┘

In [ ]:
# Joining several branches of code
